widgets/src/menu.rs

```diff
@@ -0,0 +1,754 @@
+                }
+              }
+            },
+            MenuEventData::Enter(item) => {
+              if let Some(menu) = sub_menu.as_ref() {
+                if !menu.is_show() {
+                  let id = e.current_target();
+                  item.show_sub_menu(menu, id, &wnd);
+                }
+              }
+            },
+            _ => (),
+          }
+        },
+        on_pointer_move: move |e| {
+          let item = Provider::of::<MenuItemControl>(&e).expect("menuitem must in menu");
+          item.enter(&e.window());
+        },
+        on_tap: move |e| {
+          let item = Provider::of::<MenuItemControl>(&e).expect("menuitem must in menu");
+          item.enter(&e.window());
+        },
+        @ { leading }
+        @ { label }
+        @ { trailing_text }
+        @ { trailing }
+      }
+    }
+    .into_widget()
+  }
+}
+
+/// Menu item controller, the child of Menu can visit the MenuItemControl
+/// through the provider
+#[derive(Clone)]
+pub struct MenuItemControl {
+  idx: usize,
+  menu: MenuControl,
+}
+
+impl MenuItemControl {
+  /// emit the MenuEventData::Enter event
+  pub fn enter(&self, wnd: &Sc<Window>) {
+    if self.menu.select(Some(self.idx), wnd) {
+      self.menu.enter(self.idx, wnd);
+    }
+  }
+
+  /// check if the item is selected
+  pub fn is_selected(&self) -> bool { self.menu.0.borrow().selected == Some(self.idx) }
+
+  /// check if the item is disabled
+  pub fn is_disabled(&self) -> bool { self.menu.0.borrow().items[self.idx].disabled }
+
+  /// emit the MenuEventData::Complete event, normal it will bubble up and close
+  /// all the menu. if you want to keep the menu, you should call
+  /// stop_propagation of the event
+  pub fn complete(&self, data: Option<impl Any>, wnd: &Sc<Window>) {
+    let data = data.map(|v| Sc::new_any(v));
+    self.bubble(MenuEventData::Complete(self.clone(), data.clone()), wnd);
+  }
+
+  /// close the current menu and not bubble up
+  pub fn close(&self, wnd: &Sc<Window>) { self.menu.close(wnd); }
+
+  /// disable the item, then it can't not be selected from the menu.
+  pub fn disable(&self, disabled: bool) {
+    self.menu.0.borrow_mut().items[self.idx].disabled = disabled;
+  }
+
+  /// show the sub menu around the widget
+  pub fn show_sub_menu(&self, sub_menu: &MenuControl, around_wid: WidgetId, wnd: &Sc<Window>) {
+    let pos = wnd.map_to_global(Point::zero(), around_wid);
+    let size = wnd.widget_size(around_wid).unwrap();
+    let rc = Rect::new(pos, size);
+    sub_menu.show_around(rc, Some(self.clone()), wnd);
+  }
+
+  /// return the parent item if exists(for sub menu)
+  pub fn parent(&self) -> Option<MenuItemControl> { self.menu.0.borrow().item_trigger.clone() }
+
+  fn bubble(&self, data: MenuEventData, wnd: &Sc<Window>) {
+    if let Some(from) = self.menu.0.borrow().items[self.idx].wid.get() {
+      wnd.bubble_custom_event(from, data);
+    }
+  }
+}
+
+#[derive(Template)]
+pub enum MenuChild {
+  Item(Widget<'static>),
```

Comment:
We can directly use `MenuItem` here and restrict users to only use `MenuItem` as a child.